posts/distributed-tracing-with-xray.md

```diff
@@ -0,0 +1,269 @@
+			<artifactId>aws-xray-recorder-sdk-aws-sdk-v2-instrumentor</artifactId>
+			<version>2.8.0</version>
+		</dependency>
+		<dependency>
+			<groupId>com.amazonaws</groupId>
+			<artifactId>aws-xray-recorder-sdk-spring</artifactId>
+			<version>2.8.0</version>
+		</dependency>
+```
+## Tracing incoming requests
+
+To trace all incoming requests to our microservice, we can configure a Servlet Filter provided by the SDK, we then just need to add the servlet to our spring configuration.
+
+      ``` java
+      @Configuration
+      public class AwsXRayConfig {
+
+        @Bean
+        public Filter TracingFilter() {
+          return new AWSXRayServletFilter("micro1-app");
+        }
+      }
+      ```
+
+After adding the X-Ray servlet filter to our application, the X-Ray SDK for Java creates a segment for each incoming request with an http block that contains the following information:
+
+ - HTTP method: GET, POST, PUT, DELETE, etc.
+ - Client address: The IP address of the client that sent the request.
+ - Response code: The HTTP response code for the completed request.
+ - Timing: The start time (when the request was received) and end time (when the response was sent).
+ - User agent: The user-agent from the request.
+ - Content length: The content-length from the response.
+
+## Tracing your spring bean components
+The X-Ray SDK offer a non-invasive spring integration way of tracing our spring beans which will help us see the complete call stack of the application from the controller down through the back end layers.
+ - We need to add first a spring AOP that will catch your methods execution and generate your tracing metada
+
+    ``` java
+    @Aspect
+    @Component
+    public class XRayInspector extends BaseAbstractXRayInterceptor {
+
+        @Override
+        protected Map<String, Map<String, Object>> generateMetadata(ProceedingJoinPoint proceedingJoinPoint,
+                Subsegment subsegment) {
+            return super.generateMetadata(proceedingJoinPoint, subsegment);
+        }
+
+        @Override
+        @Pointcut("@within(com.amazonaws.xray.spring.aop.XRayEnabled) && bean(*)")
+        public void xrayEnabledClasses() {
+        }
+
+    }
+    ```
+
+ - Let's annotate the spring components that we want to trace
+
+    ``` java
+    @Component
+    @XRayEnabled
+    public class HelloService { 
+      
+        public String greetingHello() {
+            return "Hello from micro 1 service";
+        }
+    }
+    ```
+
+  The @XRayEnabled annotation can be used on the class or the method level for a fine-grained control.
+
+
+## Tracing your outbound HTTP calls
+When making external calls to API's, we will need to use the AWS X-Ray Apache library which act as a wrapper to the Apache HttpComponents, this is the only time when we may have an impact on the existing code, espetially if  another HTTP client is used other than Apache like Spring Rest Template or HTTP Client (available since Java 11).
```

Comment:
```suggestion
When making external calls to API's, we will need to use the AWS X-Ray Apache library which act as a wrapper to the Apache HttpComponents. This is the only time when we may have an impact on the existing code, especially if  another HTTP client is used other than Apache like Spring Rest Template or HTTP Client (available since Java 11).
```